File: scripts/process-member-photos.py
# -*- coding: utf-8 -*-
"""Remove backgrounds from ICON photos and fit to demo avatar canvas (220x400 PNG)."""

from __future__ import annotations

import json
import sys
from io import BytesIO
from pathlib import Path

from PIL import Image
from rembg import remove

# Avoid Windows console UnicodeEncodeError on Chinese names
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
if hasattr(sys.stderr, "reconfigure"):
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

ICON_DIR = Path(r"c:\Users\heiin\Desktop\童軍管理平台\ICON")
ROOT = Path(__file__).resolve().parents[1]
OUT_DIR = ROOT / "assets" / "members"
MEMBERS_PATH = ROOT / "data" / "members.json"

CANVAS_W = 220
CANVAS_H = 400

DEMO_FILENAMES = {
    "盧羿衡": "lu-yiheng.png",
    "吳溢潼": "wu-yitong.png",
    "吳承軒": "wu-chengxuan.png",
}


def pick_icon_files(directory: Path) -> dict[str, Path]:
    by_name: dict[str, tuple[Path, int]] = {}
    for path in directory.iterdir():
        if not path.is_file():
            continue
        if path.suffix.lower() not in {".jpg", ".jpeg", ".png", ".webp"}:
            continue
        size = path.stat().st_size
        prev = by_name.get(path.stem)
        if prev is None or size > prev[1]:
            by_name[path.stem] = (path, size)
    return {name: item[0] for name, item in by_name.items()}


def out_filename(member: dict) -> str:
    name = member["name"]
    if name in DEMO_FILENAMES:
        return DEMO_FILENAMES[name]
    return f"{member['scoutId']}.png"


def fit_to_canvas(cutout: Image.Image) -> Image.Image:
    rgba = cutout.convert("RGBA")
    bbox = rgba.getbbox()
    if bbox:
        rgba = rgba.crop(bbox)

    tw, th = rgba.size
    scale = min(CANVAS_W / tw, CANVAS_H / th)
    new_w = max(1, int(round(tw * scale)))
    new_h = max(1, int(round(th * scale)))
    resized = rgba.resize((new_w, new_h), Image.Resampling.LANCZOS)

    canvas = Image.new("RGBA", (CANVAS_W, CANVAS_H), (0, 0, 0, 0))
    left = (CANVAS_W - new_w) // 2
    top = CANVAS_H - new_h  # bottom-aligned
    canvas.paste(resized, (left, top), resized)
    return canvas


def process_one(src: Path, out: Path) -> None:
    print(f"processing {src.name} -> {out.name}")
    with src.open("rb") as f:
        raw = f.read()
    cut = remove(raw)
    img = Image.open(BytesIO(cut))
    canvas = fit_to_canvas(img)
    out.parent.mkdir(parents=True, exist_ok=True)
    canvas.save(out, format="PNG", optimize=True)
    print(f"  done {canvas.size[0]}x{canvas.size[1]}")


def main() -> None:
    icons = pick_icon_files(ICON_DIR)
    data = json.loads(MEMBERS_PATH.read_text(encoding="utf-8"))
    members = data["members"] if isinstance(data, dict) else data

    updated = 0
    missing: list[str] = []
    processed: list[str] = []

    for member in members:
        name = member["name"]
        src = icons.pop(name, None)
        if src is None:
            if name not in DEMO_FILENAMES:
                missing.append(name)
            continue

        filename = out_filename(member)
        out = OUT_DIR / filename
        if out.exists():
            print(f"reuse existing {filename}")
        else:
            process_one(src, out)

        member["photo"] = f"assets/members/{filename}"
        updated += 1
        processed.append(f"{name} -> {filename}")

    MEMBERS_PATH.write_text(
        json.dumps(data, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )

    print(f"\nupdated {updated}")
    for line in processed:
        print(" ", line)
    if icons:
        print("\nunmatched ICON files:")
        for name in sorted(icons):
            print(" ", name)
    if missing:
        print("\nmembers without ICON:")
        for name in missing:
            print(" ", name)


if __name__ == "__main__":
    main()

File: scripts/process-member-photos.js
const fs = require("fs");
const path = require("path");
const sharp = require("sharp");
const { removeBackground } = require("@imgly/background-removal-node");

const ICON_DIR = "c:\\Users\\heiin\\Desktop\\童軍管理平台\\ICON";
const OUT_DIR = path.join(__dirname, "..", "assets", "members");
const MEMBERS_PATH = path.join(__dirname, "..", "data", "members.json");

const CANVAS_W = 220;
const CANVAS_H = 400;

const DEMO_FILENAMES = {
  盧羿衡: "lu-yiheng.png",
  吳溢潼: "wu-yitong.png",
  吳承軒: "wu-chengxuan.png",
};

function pickIconFiles(dir) {
  const byName = new Map();
  for (const file of fs.readdirSync(dir)) {
    const ext = path.extname(file).toLowerCase();
    if (![".jpg", ".jpeg", ".png", ".webp"].includes(ext)) continue;
    const name = path.basename(file, ext);
    const full = path.join(dir, file);
    const size = fs.statSync(full).size;
    const prev = byName.get(name);
    // Prefer larger file when duplicates exist (e.g. 胡喬立.jpg / .jpeg)
    if (!prev || size > prev.size) {
      byName.set(name, { name, full, size, ext });
    }
  }
  return byName;
}

function outFilename(member) {
  if (DEMO_FILENAMES[member.name]) return DEMO_FILENAMES[member.name];
  return `${member.scoutId}.png`;
}

async function fitToCanvas(pngBuffer) {
  const trimmed = await sharp(pngBuffer)
    .trim({ threshold: 8 })
    .ensureAlpha()
    .png()
    .toBuffer();

  const meta = await sharp(trimmed).metadata();
  const maxW = CANVAS_W;
  const maxH = CANVAS_H;
  const fitScale = Math.min(maxW / meta.width, maxH / meta.height);
  const tw = Math.max(1, Math.round(meta.width * fitScale));
  const th = Math.max(1, Math.round(meta.height * fitScale));

  const resized = await sharp(trimmed)
    .resize(tw, th, { fit: "inside", withoutEnlargement: false })
    .png()
    .toBuffer();

  const left = Math.round((CANVAS_W - tw) / 2);
  const top = CANVAS_H - th; // bottom-aligned like CSS object-position: center bottom

  return sharp({
    create: {
      width: CANVAS_W,
      height: CANVAS_H,
      channels: 4,
      background: { r: 0, g: 0, b: 0, alpha: 0 },
    },
  })
    .composite([{ input: resized, left, top }])
    .png()
    .toBuffer();
}

async function processOne(srcPath, outPath) {
  console.log("processing", path.basename(srcPath), "->", path.basename(outPath));
  // Normalize to PNG buffer/blob so imgly can decode reliably on Windows
  const pngBuffer = await sharp(srcPath).rotate().png().toBuffer();
  const inputBlob = new Blob([pngBuffer], { type: "image/png" });
  const blob = await removeBackground(inputBlob, {
    model: "medium",
    output: { format: "image/png", quality: 0.9 },
  });
  const cutout = Buffer.from(await blob.arrayBuffer());
  const canvas = await fitToCanvas(cutout);
  fs.writeFileSync(outPath, canvas);
  const meta = await sharp(canvas).metadata();
  console.log("  done", meta.width, "x", meta.height, "alpha=", meta.hasAlpha);
}

async function main() {
  fs.mkdirSync(OUT_DIR, { recursive: true });
  const icons = pickIconFiles(ICON_DIR);
  const data = JSON.parse(fs.readFileSync(MEMBERS_PATH, "utf8"));
  const members = data.members || data;

  let updated = 0;
  const missing = [];
  const processed = [];

  for (const member of members) {
    const icon = icons.get(member.name);
    if (!icon) {
      if (!DEMO_FILENAMES[member.name]) missing.push(member.name);
      continue;
    }

    const filename = outFilename(member);
    const outPath = path.join(OUT_DIR, filename);
    await processOne(icon.full, outPath);
    member.photo = `assets/members/${filename}`;
    updated++;
    processed.push(`${member.name} -> ${filename}`);
    icons.delete(member.name);
  }

  fs.writeFileSync(MEMBERS_PATH, JSON.stringify(data, null, 2) + "\n");

  console.log("\nupdated", updated);
  processed.forEach((l) => console.log(" ", l));
  if (icons.size) {
    console.log("\nunmatched ICON files:");
    for (const [name] of icons) console.log(" ", name);
  }
  if (missing.length) {
    console.log("\nmembers without ICON:");
    missing.forEach((n) => console.log(" ", n));
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
